### PR TITLE
Extract CLI logic from `clinic.py`

### PR DESCRIPTION
**Pull request by AlexWaygood**
- Move `Tools/clinic/clinic.py` and `Tools/clinic/cpp.py` inside the `Tools/clinic/libclinic` package.
- Extract the CLI logic from `clinic.py` into a new `libclinic` submodule, `libclinic.cli`
- Add a new file, `Tools/clinic/run_clinic.py`, which simply imports `libclinic.cli` and executes the CLI if `__name__ == "__main__"`. Update the Makefile etc. so that they run `Tools/clinic/run_clinic.py` instead of `Tools/clinic/clinic.py`.

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Nice

File: Tools/clinic/libclinic/cli.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+from typing import NoReturn
+
+from libclinic import clinic
+
+__all__ = ["main"]
+
+
+def create_cli() -> argparse.ArgumentParser:
+    cmdline = argparse.ArgumentParser(
+        prog="Argument Clinic",
```
Comment (AlexWaygood) [inline Tools/clinic/libclinic/cli.py:16]: The failing test is correctly failing. I changed this value from `clinic.py` to `Argument Clinic`, and now we have a bad `usage` line in the `--help` message:

```
>python Tools/clinic/run_clinic.py --help
Running Debug|x64 interpreter...
usage: Argument Clinic [-h] [-f] [-o OUTPUT] [-v] [--converters] [--make] [--srcdir SRCDIR] [--exclude EXCLUDE] [--limited] [FILE ...]

Preprocessor for CPython C files. The purpose of the Argument Clinic... (truncated)
```

The easiest fix is just to hardcode `run_clinic.py` in here:

```suggestion
        prog="Tools/clinic/run_clinic.py",
```

...but it's slightly icky for code in the `libclinic` package to know the name of the file calling it. Another possible fix is to move all the CLI logic into `run_clinic.py`, but I kinda want the CLI logic to be a `libclinic` submodule.

...Thoughts?

Comment (erlend-aasland) [reply Tools/clinic/libclinic/cli.py:16]: How about this (diff relative to PR):

```diff
diff --git a/Lib/test/test_clinic.py b/Lib/test/test_clinic.py
index 2bcb3a5c73..1ee8dce02d 100644
--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -2590,7 +2590,7 @@ def test_cli_verbose(self):
 
     def test_cli_help(self):
         out = self.expect_success("-h")
-        self.assertIn("usage: clinic.py", out)
+        self.assertIn("Preprocessor for CPython C files.", out)
 
     def test_cli_converters(self):
         prelude = dedent("""
diff --git a/Tools/clinic/libclinic/cli.py b/Tools/clinic/libclinic/cli.py
index 7242eea0bc..1cc0c46fe7 100644
--- a/Tools/clinic/libclinic/cli.py
+++ b/Tools/clinic/libclinic/cli.py
@@ -13,7 +13,6 @@
 
 def create_cli() -> argparse.ArgumentParser:
     cmdline = argparse.ArgumentParser(
-        prog="Argument Clinic",
         description="""Preprocessor for CPython C files.
 
 The purpose of the Argument Clinic is automating all the boilerplate involved
```

ArgumentParser uses the basename of `sys.argv[0]` by default:

```console
$ python3 Tools/clinic/run_clinic.py -h | head -n 1
usage: run_clinic.py [-h] [-f] [-o OUTPUT] [-v] [--converters] [--make] [--srcdir SRCDIR] [--exclude EXCLUDE] [--limited] [FILE ...]
```

Comment (erlend-aasland) [reply Tools/clinic/libclinic/cli.py:16]: (The point of the help check, as I see it, is to see that `argparse` was invoked and our customised help message returned. IMO, it does not matter much to check the name of the program. The description is a better validation.)

Comment (AlexWaygood) [reply Tools/clinic/libclinic/cli.py:16]: > ArgumentParser uses the basename of `sys.argv[0]` by default:

Oh nice, that solves things! I'll make the change when I'm back home 👍

**Review by erlend-aasland — APPROVED** [review]

**Comment by AlexWaygood** [comment]
(I'll probably not get back to this until 27th now — will upstream this then :)

**Comment by erlend-aasland** [comment]
> (I'll probably not get back to this until 27th now — will upstream this then :)

Enjoy the days off 🎅🎄🎉

**Comment by erlend-aasland** [comment]
Hm, I think `parse_file()` should live where the CLI lives; it is an app function, not a library function.

**Review by erlend-aasland — APPROVED** [review]

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request closed without merging**